src/algebra/monoid_algebra.lean

```diff
@@ -632,6 +698,45 @@ f.single_mul_apply_aux r _ _ _ $ λ a, by rw [zero_add]
 
 end misc_theorems
 
+end add_monoid_algebra
+
+/-!
+#### Conversions between `add_monoid_algebra` and `monoid_algebra`
+
+While we were not able to define `add_monoid_algebra k G = monoid_algebra k (multiplicative G)` due
+to definitional inconveniences, we can still show the types are isomorphic.
+-/
+
+/-- The equivalence between `add_monoid_algebra` and `monoid_algebra` in terms of `multiplicative` -/
+protected def add_monoid_algebra.to_multiplicative [semiring k] [add_monoid G] :
+  add_monoid_algebra k G ≃+* monoid_algebra k (multiplicative G) :=
+{ map_mul' := λ x y, by {
+    simp only [finsupp.dom_congr],
+    apply add_monoid_algebra.map_domain_mul',
+    intros p q,
+    -- At this point the pretty-printing of the goal is nonsense, containing `p + q` despite
+    -- the fact there is no obvious `has_add (multiplicative G)`.
```

Comment:
This is because you abused type equality when you applied `map_domain_mul'` -- it was an act of desperation by the unifier. If `multiplicative` was irreducible, which it should be in my opinion, then people would find this much harder to do and they would be forced to write better (in the sense of type correct) code.